src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -152,138 +153,185 @@ pub struct WorkunitStore {
 
 #[derive(Clone)]
 struct StreamingWorkunitData {
-  started_workunits_rx: Arc<Mutex<Receiver<Workunit>>>,
-  started_workunits_tx: Arc<Mutex<Sender<Workunit>>>,
-  completed_workunits_rx: CompletedWorkunitReceiver,
-  completed_workunits_tx: CompletedWorkunitSender,
+  msg_rx: Arc<Mutex<Receiver<StoreMsg>>>,
+  msg_tx: Arc<Mutex<Sender<StoreMsg>>>,
   workunit_records: Arc<Mutex<HashMap<SpanId, Workunit>>>,
 }
 
 impl StreamingWorkunitData {
   fn new() -> StreamingWorkunitData {
-    let (started_workunits_tx, started_workunits_rx) = channel();
-    let (completed_workunits_tx, completed_workunits_rx) = channel();
+    let (msg_tx, msg_rx) = channel();
     StreamingWorkunitData {
-      started_workunits_tx: Arc::new(Mutex::new(started_workunits_tx)),
-      started_workunits_rx: Arc::new(Mutex::new(started_workunits_rx)),
-      completed_workunits_tx: Arc::new(Mutex::new(completed_workunits_tx)),
-      completed_workunits_rx: Arc::new(Mutex::new(completed_workunits_rx)),
+      msg_tx: Arc::new(Mutex::new(msg_tx)),
+      msg_rx: Arc::new(Mutex::new(msg_rx)),
       workunit_records: Arc::new(Mutex::new(HashMap::new())),
     }
   }
+
+  pub fn with_latest_workunits<F, T>(&mut self, max_verbosity: log::Level, f: F) -> T
+  where
+    F: FnOnce(&[Workunit], &[Workunit]) -> T,
+  {
+    use std::collections::hash_map::Entry;
+    let should_emit = |workunit: &Workunit| -> bool { workunit.metadata.level <= max_verbosity };
+
+    let (started_workunits, completed_workunits) = {
+      let mut workunit_records = self.workunit_records.lock();
+
+      let mut started_messages = vec![];
+      let mut completed_messages = vec![];
+
+      {
```

Comment:
It doesn't look like the lock on `self.workunit_records` needs to be acquired while `self.msg_rx` is acquired here.